### PR TITLE
update spa to ~eamv3b2 and fix unit inconsistency in spa

### PR DESCRIPTION
**Pull request by mahf708**
xref #2608  

todo:
- [x] throw a runtime error if wrong units are provided in input files
- [ ] some test guards
- [x] ~non-bfb procedure~ answer may vary a little due to offline/online density, but not important
- [x] need to update input file

### CODE REVIEW

File: components/eam/src/physics/p3/eam/micro_p3.F90 [hunk 1]
```diff
@@ -503,7 +503,10 @@ SUBROUTINE p3_main_part1(kts, kte, kbot, ktop, kdir, do_predict_nc, do_prescribe
       !    prescribe that value
 
           if (do_prescribed_CCN) then
-             nc(k) = max(nc(k),nccn_prescribed(k))
+             ! nccn_prescribed is in #/cm3, so needs to be converted
+             ! *1e6: #/cm3 --> #/m3
+             ! *inv_rho: #/m3 --> #/kg 
+             nc(k) = max(nc(k),nccn_prescribed(k)*1.0e6_rtype*inv_rho(k))
```
[the remark this answers is not in the document]
Comment (PeterCaldwell) [reply components/eam/src/physics/p3/eam/micro_p3.F90:509]: I love the code comments you added describing the units conversion. I also like your git comment about CCN being the same value in cloud and outside, so no need to scale by cloud fraction (except to undo what P3 is going to do). Maybe add this as a code comment? But also - It looks like you *aren't* multiplying by cloud fraction here even though you said you need to. Is this intended?

File: components/eam/src/physics/p3/scream/micro_p3_interface.F90 [hunk 2]
```diff
@@ -1069,8 +1069,6 @@ subroutine micro_p3_tend(state, ptend, dtime, pbuf)
 
     if (do_prescribed_CCN) then
        call pbuf_get_field(pbuf, ccn3_idx, ccn_trcdat) ! now you can use ccn_trcdat anywhere in this code
-       ! ccn is uniformly distributed throughout the cell, but P3 computes in-cloud values assuming cell-averages are comprised
-       ! of zero values outside cloud. Preemptively multiplying by cldfrac here is needed to get the correct in-cloud ccn value in P3
```
Comment (PeterCaldwell) [inline components/eam/src/physics/p3/scream/micro_p3_interface.F90:1073]: Again, nice comment but not actually implemented?

File: components/eamxx/src/physics/p3/impl/p3_main_impl_part1.hpp [hunk 3]
```diff
@@ -131,7 +131,10 @@ ::p3_main_part1(
       // prescribe that value
 
       if (do_prescribed_CCN) {
-         nc(k).set(not_drymass, max(nc(k), nccn_prescribed(k)/inv_cld_frac_l(k)));
```
Comment (PeterCaldwell) [inline components/eamxx/src/physics/p3/impl/p3_main_impl_part1.hpp:134]: isn't div by inv_cld_frac just multiplying by cldfrac? :-)

File: components/eam/src/physics/p3/scream/micro_p3.F90 [hunk 4]
```diff
@@ -410,8 +410,14 @@ SUBROUTINE p3_main_part1(kts, kte, kbot, ktop, kdir, do_predict_nc, do_prescribe
...
-             !nccn_prescribed is an in-cloud value so make it grid average in this assignment
-             nc(k) = max(nc(k),nccn_prescribed(k)/inv_cld_frac_l(k))
+             ! nccn_prescribed is in #/cm3, so needs to be converted
+             ! *1e6: #/cm3 --> #/m3
+             ! *inv_rho: #/m3 --> #/kg 
+             ! /inv_cld_frac_l: to guard against unrealistiically inflating nc inside P3 ---
+             !                  CCN is in-grid, but realistically it is not related to 
+             !                  "in-cloud" properties, thus to retain its value everywhere,
+             !                  multiply here by the cloud fraction to avoid inflating it inside P3
+             nc(k) = max(nc(k),nccn_prescribed(k)*1.0e6_rtype*inv_rho(k)/inv_cld_frac_l(k))
```
Comment (PeterCaldwell) [inline components/eam/src/physics/p3/scream/micro_p3.F90:420]: Same comment as above.

File: components/eamxx/src/physics/p3/impl/p3_main_impl_part1.hpp [hunk 5]
```diff
@@ -131,7 +131,14 @@ ::p3_main_part1(
...
+         // nccn_prescribed is in #/cm3, so needs to be converted
+         // *1e6: #/cm3 --> #/m3
+         // *inv_rho: #/m3 --> #/kg 
+         // /inv_cld_frac_l: to guard against unrealistiically inflating nc inside P3 ---
+         //                  CCN is in-grid, but realistically it is not related to 
+         //                  "in-cloud" properties, thus to retain its value everywhere,
+         //                  multiply here by the cloud fraction to avoid inflating it inside P3
+         nc(k).set(not_drymass, max(nc(k), nccn_prescribed(k)*1e6*inv_rho(k)/inv_cld_frac_l(k)));
```
Comment (PeterCaldwell) [inline components/eamxx/src/physics/p3/impl/p3_main_impl_part1.hpp:141]: same as above

**Comment by bartgol** [comment]
I have no idea what that error means. You should ask the Scorpio folks, they may be able to explain the cause..

**Pull request closed without merging**